### DISCUSSION TITLE
prisma-multi-tenant missing prisma2 cli

### DISCUSSION THREAD

**Discussion by BjoernRave** [vercel/vercel#3900] [General]
Hey, so there is this library which uses a cli tool for parts of its functionality. When deploying my app to Zeit Now the cli tool doesn't get included in the package though, creating problems in prod.

 Is there a way to explicitly include an npm package into the bundle with something like `includeFiles`?

I tried around a bit with `includeFiles`, but couldn't get it to work.

The package I am talking about is:
 https://github.com/Errorname/prisma-multi-tenant, which is using https://github.com/prisma/prisma2 to programmatically create new tenants

@leo

**Comment by styfle** [comment]
Hi @BjoernRave 

I would recommend using the programatic UI instead of relying on the CLI in your Serverless Function.

You can probably do `const { MultiTenant } = require('prisma-multi-tenant')`.

**Reply by BjoernRave** [reply]
thanks for the info. The problem is, that `prisma-multi-tenant` uses the prisma2 cli under the hood to create new tenants. So the programmatic UI of `prisma-multi-tenant` doesn't work as expected on Zeit Now, because the `prisma2` CLI tools are not available. I think the reason `prisma-multi-tenant` uses the cli, is because the `prisma2` sdk is still WIP as well as the `migrate` package.
 @Errorname can provide more context if needed

**Reply by janpio** [reply]
Hey @BjoernRave, can you elaborate exactly what is failing because of the missing Prisma2 CLI? My knowledge of `prisma-multi-tenant` is quite limited unfortunately. What "action" of `prisma-multi-tenant` is using the CLI how exactly here? Maybe we can help figure out a way around that (or at least understand better).

**Reply by BjoernRave** [reply]
Hey @janpio, creating new tenants is failing in prod. the command is executing a `prisma2 migrate up --create-db --experimental`, to create a new database for the tenant and migrate it.  prisma-multi-tenant uses the cli for that, because there is no sdk yet, @Errorname tried to use the `migrate` directly, but this was his response to that: 

>I tried bypassing the prisma2 migrate command and use the @prisma/migrate package, but I can’t because of a lot of side-effects happening when downloading the command


The error I get when performing the action is: 

```
Command failed: "prisma2" migrate up --create-db --experimental /bin/sh: prisma2: command not found
```

**Reply by williamluke4** [reply]
@BjoernRave Have a look at https://github.com/zeit/node-file-trace/issues/104#issuecomment-598424374

**Reply by janpio** [reply]
Thanks @BjoernRave - but what command are _you_ executing, and how? Currently I can not follow what you are actually doing, which would be helpful to understand the whole thing.

**Reply by BjoernRave** [reply]
I am basically executing this code: 
```
import { MultiTenant } from 'prisma-multi-tenant'

const multiTenant = new MultiTenant()

await multiTenant.createTenant({
  name: 'test',
          provider: 'postgresql',
          url: 'postgresql://postgres...',
})
```

That code is executing this under the hood: `prisma2 migrate up --create-db --experimental`

I think prisma multi tenant is setting env vars accordingly, so the command get's executed on the new tenanat db.

Hope, that's what you are looking for

**Reply by janpio** [reply]
That is exactly what I was looking for. Thanks, I can investigate from here then. 

(If you of course have [or want to create] a simple reproduction repository, that I can just check out and play with, that would make me very happy.)

**Reply by janpio** [reply]
(A colleague should have reached out to you via Slack, so we can figure out what exactly is going wrong here - and hopefully have a solution soon.)

**Reply by steebchen** [reply]
From an internal discussion between @BjoernRave, @Errorname and me:

To clarify, the `prisma2` command is not available in the production zeit environment, since zeit's package mechanism doesn't pick up the `prisma2` binary, which is expected.

The correct way to solve this is to offer all methods via @prisma/sdk or @prisma/migrate where Prisma handles everything, but unfortunately we don't have the capacity for that at this moment. This means that we need to find a workaround for now.

We came to the conclusion that for now, your best bet is to avoid spawning prisma2 CLI commands in production, and instead dig through our migrate javascript code (there's no docs yet) and invoke these methods directly. 

One problem you might run into is that the migration engine is not available in production. If you run into this, please let us know and we'll look into it. There might need to be some adaptions so that the zeit/now bundler will pick up the migration engine binary, and we may need to make sure that the migration engine actually exists when it is required in production.

**Comment by williamluke4** [comment]
Hey @styfle, I'm having a very similar issue using `@prisma/sdk` on a nextjs API route. 
Please see the error below from the api route. 
```
22:19:30:76
2020-03-12T19:19:30.792Z	5f8000e1-e33b-435e-93ff-066289995b7d	ERROR	Error: Could not find query-engine binary. Searched in /var/task/node_modules/@prisma/sdk/query-engine-rhel-openssl-1.0.x and /var/task/node_modules/@prisma/query-engine-rhel-openssl-1.0.x
    at /var/task/node_modules/@prisma/sdk/dist/engineCommands.js:61:15
    at Generator.next (<anonymous>)
    at fulfilled (/var/task/node_modules/@prisma/sdk/dist/engineCommands.js:5:58)
```

**Reply by styfle** [reply]
I just tried the latest `@prisma/sdk`, currently `2.0.0-preview024` and it worked for me.

```js
const sdk = require('@prisma/sdk');
const found = !!sdk; // true if found
module.exports = (req, res) => res.json({ found });
```

Can you provide example code that is failing?

**Reply by williamluke4** [reply]
This is the api route that is failing 
```tsx
import { getDMMF } from '@prisma/sdk';
import { NextApiRequest, NextApiResponse } from 'next';

const defaultSchema = `
generator photon {
  provider = "photonjs"
}

model Test {
  id                 String           @default(cuid()) @id
  createdAt          DateTime         @default(now())
  updatedAt          DateTime         @updatedAt
  firstName          String
  lastName           String
  email              String           @unique
  password           String
  phone              String
  responseRate       Float?
  responseTime       Int?
}
`;

export default async (req: NextApiRequest, res: NextApiResponse) => {
  const data = req.body as string;
  res.statusCode = 200;
  const dmmf = await getDMMF({ datamodel: data || defaultSchema});
  res.json(dmmf);
};
```

**Reply by styfle** [reply]
Thanks I was able to reproduce!

I created an issue here: https://github.com/zeit/node-file-trace/issues/104
